staging/src/k8s.io/externaljwt/pkg/plugin/plugin.go

```diff
@@ -0,0 +1,263 @@
+}
+
+// GetTokenVerificationKeys returns a map of supported external keyIDs to keys
+// the keys are PKIX-serialized.
+func (p *Plugin) GetTokenVerificationKeys(ctx context.Context) (*VerificationKeys, int, error) {
+	ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
+
+	req := &ejspb.FetchKeysRequest{}
+	resp, err := p.client.FetchKeys(ctx, req)
+	if err != nil {
+		return nil, 0, fmt.Errorf("while getting externally supported jwt signing keys: %w", err)
+	}
+
+	keys := []serviceaccount.PublicKey{}
+	for _, protoKey := range resp.GetKeys() {
+		parsedPublicKey, err := x509.ParsePKIXPublicKey(protoKey.GetKey())
+		if err != nil {
+			return nil, 0, fmt.Errorf("while parsing external public keys: %w", err)
+		}
+
+		keys = append(keys, serviceaccount.PublicKey{
+			KeyID:                    protoKey.GetKeyId(),
+			PublicKey:                parsedPublicKey,
+			ExcludeFromOidcDiscovery: protoKey.ExcludeFromOidcDiscovery,
+		})
+	}
+
+	vk := &VerificationKeys{
+		Keys:          keys,
+		DataTimestamp: resp.GetDataTimestamp().AsTime(),
+	}
+
+	return vk, int(resp.RefreshHintSeconds), nil
+}
+
+// GetServiceMetadata returns metadata associated with externalJWTSigner
+// It Includes details like max token lifetime supported by externalJWTSigner, etc.
+func (p *Plugin) GetServiceMetadata(ctx context.Context) (*Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	req := &ejspb.MetadataRequest{}
+	resp, err := p.client.Metadata(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("while getting metadata from external jwt signer: %w", err)
+	}
+
+	return &Metadata{
+		MaxTokenExpirationSeconds: int(resp.MaxTokenExpirationSeconds),
+	}, nil
+}
+
+// IsExternal informs if TokenGenerator intigerates with external-jwt-signer.
+func (p *Plugin) IsExternal() bool {
+	return true
+}
+
+// setKeyCache is for use with testcases.
+// It is required because creating keyCache itself requires a reference to Plugin.
+func (p *Plugin) setKeyCache(keyCache *KeyCache) {
+	p.keyCache = keyCache
+}
+
+func mergeClaims(claims *jwt.Claims, privateClaims interface{}) ([]byte, error) {
+	privateBytes, err := json.Marshal(privateClaims)
+	if err != nil {
+		return nil, fmt.Errorf("while marshalling private claims: %w", err)
+	}
+
+	var unifiedMap map[string]any
+	if err := json.Unmarshal(privateBytes, &unifiedMap); err != nil {
+		return nil, fmt.Errorf("while unmarshalling private claims: %w", err)
+	}
+
+	unifiedMap["iss"] = claims.Issuer
+	unifiedMap["sub"] = claims.Subject
+	unifiedMap["aud"] = claims.Audience
+	unifiedMap["exp"] = claims.Expiry
+	unifiedMap["nbf"] = claims.NotBefore
+	unifiedMap["iat"] = claims.IssuedAt
+	if claims.ID != "" {
+		unifiedMap["jti"] = claims.ID
+	}
+
+	unifiedBytes, err := json.Marshal(unifiedMap)
+	if err != nil {
+		return nil, fmt.Errorf("while marshalling unified claims: %w", err)
+	}
+
+	return unifiedBytes, nil
```

Comment:
return map[string]interface{}, so the return value can also be used by the standard signer, move the json.Marshal / base64-encode into the external-signer use

Reply:
this comment isn't relevant if we use jose to do the payload construction / merging

Reply:
Done